server/websocket_test.go

```diff
@@ -960,4 +961,121 @@ func TestHandleJoin(t *testing.T) {
+
+		mockAPI.On("GetUsers", mock.AnythingOfType("*model.UserGetOptions")).Return([]*model.User{
+			{
+				Id:     "adminID",
+				Locale: "it",
+			},
+		}, nil).Once()
+
+		mockAPI.On("GetDirectChannel", "adminID", "").Return(&model.Channel{
+			Id: "channelID",
+		}, nil).Once()
+
+		mockAPI.On("IsEnterpriseReady").Return(false).Once()
+
+		mockAPI.On("CreatePost", &model.Post{
+			UserId:    "",
+			ChannelId: "channelID",
+			Message:   ":warning: app.admin.concurrent_sessions_warning.intro\r\n\r\napp.admin.concurrent_sessions_warning.team",
+		}).Return(&model.Post{Id: "postID"}, nil).Once()
+
+		mockRTCMetrics.On("IncRTCSessions", "default").Once()
+
+		mockMetrics.On("IncWebSocketEvent", "out", wsEventJoin).Once()
+		mockAPI.On("PublishWebSocketEvent", wsEventJoin, map[string]any{"connID": connID},
+			&model.WebsocketBroadcast{UserId: userID, ReliableClusterSend: true}).Once()
+
+		// DEPRECATED
+		mockMetrics.On("IncWebSocketEvent", "out", wsEventUserConnected).Once()
+		mockAPI.On("PublishWebSocketEvent", wsEventUserConnected, map[string]any{"userID": userID},
+			&model.WebsocketBroadcast{ChannelId: channelID, ReliableClusterSend: true}).Once()
+
+		mockMetrics.On("IncWebSocketEvent", "out", wsEventUserJoined).Once()
+		mockAPI.On("PublishWebSocketEvent", wsEventUserJoined, map[string]any{"session_id": connID, "user_id": userID},
+			&model.WebsocketBroadcast{ChannelId: channelID, ReliableClusterSend: true}).Once()
+
+		mockMetrics.On("IncWebSocketEvent", "out", wsEventCallState).Once()
+		mockAPI.On("PublishWebSocketEvent", wsEventCallState, mock.Anything,
+			&model.WebsocketBroadcast{UserId: userID, ReliableClusterSend: true}).Once()
+
+		mockMetrics.On("IncWebSocketConn").Once()
+
+		// Call unlock
+		mockAPI.On("KVDelete", "mutex_call_"+channelID).Return(nil).Once()
+
+		mockMetrics.On("DecWebSocketConn").Once()
+		mockRTCMetrics.On("DecRTCSessions", "default").Once()
+		mockRTCMetrics.On("IncRTCConnState", "closed").Once()
+
+		// DEPRECATED
+		mockMetrics.On("IncWebSocketEvent", "out", wsEventUserDisconnected).Once()
+		mockAPI.On("PublishWebSocketEvent", wsEventUserDisconnected, map[string]any{"userID": userID},
+			&model.WebsocketBroadcast{ChannelId: channelID, ReliableClusterSend: true}).Once()
+
+		mockMetrics.On("IncWebSocketEvent", "out", wsEventUserLeft).Once()
+		mockAPI.On("PublishWebSocketEvent", wsEventUserLeft, map[string]any{"session_id": connID, "user_id": userID},
+			&model.WebsocketBroadcast{ChannelId: channelID, ReliableClusterSend: true}).Once()
+
+		mockAPI.On("UpdatePost", mock.AnythingOfType("*model.Post")).Return(&model.Post{Id: postID}, nil).Once()
+
+		// Call unlock
+		mockAPI.On("KVDelete", "mutex_call_"+channelID).Return(nil).Once()
+
+		err := p.handleJoin(userID, connID, authSessionID, callsJoinData{
+			CallsClientJoinData: CallsClientJoinData{
+				ChannelID: channelID,
+			},
+		})
+		require.NoError(t, err)
+
```

Comment:
wow... that's a lot of mocks. how long did that take?

Reply:
A bit. The problem is that after implementing this, I realized the single subcases were dependent on the order, so they would fail if executed independently (e.g., `TestHandleJoin/admin_warning`); this is because there isn't an easy way I could find to reset mock registrations between new test cases without re-initializing them completely.

We could, in theory, try to extract all the common ones and put them above before any case starts, but then it would probably be even less clear what's happening.